service/controller/v18/resource/nodeindexstatus/create.go

```diff
@@ -0,0 +1,130 @@
+package nodeindexstatus
+
+import (
+	"context"
+	"reflect"
+	"sort"
+
+	"github.com/giantswarm/microerror"
+	"github.com/giantswarm/operatorkit/controller/context/reconciliationcanceledcontext"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/giantswarm/kvm-operator/service/controller/v18/key"
+)
+
+func (r *Resource) EnsureCreated(ctx context.Context, obj interface{}) error {
+	cr, err := key.ToCustomObject(obj)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	var nodeIndexes map[string]int
+	{
+		var idx int
+		allocations := key.AllocatedNodeIndexes(cr)
+		nodes := key.AllNodes(cr)
+		nodeIndexes = cr.Status.KVM.NodeIndexes
+		if nodeIndexes == nil {
+			nodeIndexes = make(map[string]int)
+		}
+
+		// Remove node indexes for non-existent nodes.
+		for nodeID, _ := range nodeIndexes {
+			found := false
+
+			for _, n := range nodes {
+				if n.ID == nodeID {
+					found = true
+					break
+				}
+			}
+
+			if !found {
+				delete(nodeIndexes, nodeID)
+			}
+		}
+
+		// Ensure all present nodes have node index allocation.
+		for _, n := range nodes {
+			_, exists := nodeIndexes[n.ID]
+			if !exists {
+				allocations, idx = allocateIndex(allocations)
+				nodeIndexes[n.ID] = idx
+			}
+		}
+	}
+
+	{
+		r.logger.LogCtx(ctx, "level", "debug", "message", "updating status with node indexes")
+
+		newObj, err := r.g8sClient.ProviderV1alpha1().KVMConfigs(cr.GetNamespace()).Get(cr.GetName(), metav1.GetOptions{})
```

Comment:
Shouldn't that go into the `if !reflect.DeepEqual(newObj.Status.KVM.NodeIndexes, nodeIndexes)` ?

Reply:
What do you mean here?

Reply:
Does it make sense to get the object if we don't need to update it? Getting the object is just for the latest resource ID

Reply:
Ah, right, not necessarily. I'll change this.